Support multi-index LOOKUP JOIN and various bug fixes

craigtaverner opened the pull request:
While working on supporting multi-index LOOKUP JOIN; various bugs were fixed:

* Previously we just used '*' for lookup-join indices, because the fieldnames were sometimes not being correctly determined. The problem was with KEEP referencing fields from the right that had previously been defined on the left as aliases, including using the ROW command. We normally don't want to ask for aliases, but if they could be shadowed by a lookup join, we need to keep them.
* With both single and multi-index LOOKUP JOIN we need to mark each index as potentially wildcard fields, if the KEEP commands occur before the LOOKUP JOIN.

craigtaverner committed:
Fix wildcard field-caps for LOOKUP JOIN

Previously we just used '*' for lookup-join indices, because the fieldnames were sometimes not being correctly determined. The problem was with KEEP referencing fields from the right that had previously been defined on the left as aliases, including using the ROW command. We normally don't want to ask for aliases, but if they could be shadowed by a lookup join, we need to keep them.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

alex-spies reviewed (commented):
Heya, I think this should improve things, but I think we need to add tests to really see what fieldnames we'll request. There are [tests for the field names method](https://github.com/elastic/elasticsearch/blob/85d3362ff9c728aaf4b367ebd9a4abceee7cf21a/x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/session/IndexResolverFieldNamesTests.java) and I think we should throw a bunch of LOOKUP JOIN queries in there to ensure that we're asking for the right fields. In particular, there may be bugs in situations where there's a KEEP/DROP _before_ (i.e. upstream from) the LOOKUP JOIN, which would indicate that we only need some fields from the main index - that could end up requesting too few fields from the LOOKUP JOIN.

I tried the following test, and it passes as-is on main:

```java
    public void testLookupJoin() {
        assertFieldNames(
            "FROM employees | KEEP languages | RENAME languages AS language_code | LOOKUP JOIN languages_lookup ON language_code",
            Set.of("languages", "language_code", "language_code.*", "languages.*")
        );
    }
```
Not sure that's correct - we actually need all the fields from LOOKUP JOIN.

Note: Since we supply a single list of field names that doesn't differentiate between the main index and lookup indices (c.f. [here](https://github.com/elastic/elasticsearch/blob/85d3362ff9c728aaf4b367ebd9a4abceee7cf21a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java#L314-L318) where the fieldnames are passed to preAnalyzeLookupIndices), we'll ask for all the main index' field names, too, and will request too many fields either from the main index, or from lookup indices in case of multiple lookups that shadow.

Requesting too many fields is likely okay for now as long as it's not `*`. We probably want an optimization rule that prunes the lookup to only fetch the relevant fields, and it shouldn't/can't be the job of the pre-analysis stage to determine this.

astefan reviewed (changes requested):
This may not be related to this specific PR's changes, it may be something that was postponned but, nevertheless, I think we need tests in IndexResolverFieldNamesTests for queries that use `lookup join` queries so that the `fieldNames` output is correct given those queries. This _may_ be the right PR to add the missing tests.

[The meta issue](https://github.com/elastic/elasticsearch/issues/116208) also mentions a bug related to `fieldNames`. I am not sure if what this PR is doing has any relation to that bug, but the lack of any `lookup join` related tests in `IndexResolverFieldNamesTests` is concerning.

craigtaverner commented:
> This may not be related to this specific PR's changes, it may be something that was postponned but, nevertheless, I think we need tests in IndexResolverFieldNamesTests for queries that use `lookup join` queries so that the `fieldNames` output is correct given those queries. This _may_ be the right PR to add the missing tests.

I could investigate that. The current PR replaces a fix that was already tested in the csv-spec tests, with a new fix. I did verify that removing "*" without also fixing the alias check causes failures.

> [The meta issue](https://github.com/elastic/elasticsearch/issues/116208) also mentions a bug related to `fieldNames`. I am not sure if what this PR is doing has any relation to that bug, but the lack of any `lookup join` related tests in `IndexResolverFieldNamesTests` is concerning.

The bug in the meta-issue was added by me, so yes, this is the fix for that bug. The text of the bug is "EsqlSession.fieldNames does not handle lookup references that are also mentioned in aliases (erases them)", but no specific issue was created, since it is a refactoring mainly.

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Add support for multiple lookup joins

And also improve the fix for wildcard lookups, since the previous fix did not consider KEEP before lookup-join

craigtaverner committed:
Ignore two remaining failures

craigtaverner committed:
Revert some formatting changes

craigtaverner committed:
Fix compile errors

craigtaverner committed:
Merge branch 'fix-lookupjoin-wildcard-fieldcaps' of github.com:craigtaverner/elasticsearch into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
More consistent test naming

craigtaverner committed:
More consistent test naming

craigtaverner committed:
Change EsqlCapability for backwards compatibility tests

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Add unit tests for fieldNames and LOOKUP JOIN

craigtaverner committed:
Merge branch 'fix-lookupjoin-wildcard-fieldcaps' of github.com:craigtaverner/elasticsearch into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Fix failing tests

craigtaverner committed:
Merge branch 'fix-lookupjoin-wildcard-fieldcaps' of github.com:craigtaverner/elasticsearch into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
```diff
@@ -325,3 +360,150 @@ count:long | type:keyword
+lookupIPAndMessageFromKeepRow
+required_capability: join_lookup_v5
+
+ROW left = "left", client_ip = "172.21.0.5", message = "Connected to 10.1.0.1", right = "right"
+| KEEP left, client_ip, message, right
+| LOOKUP JOIN clientips_lookup ON client_ip
+| LOOKUP JOIN message_types_lookup ON message
```
I think we could use a test like this, but with the `KEEP` between the two `LOOKUP JOIN`s. Because that could be an edge case for our fieldNames method and related fieldCaps calls.

craigtaverner replied:
I did add one to the unit tests, with KEEP in between, but could have one in the IT too.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -554,10 +517,24 @@ private static void analyzeAndMaybeRetry(
+                .collect(Collectors.toSet());
+            // get the field names from the parsed plan combined with the ENRICH match fields from the ENRICH policy
+            l.onResponse(fieldNames(parsed, enrichMatchFields, new ListenerResult(enrichResolution)));
+        } catch (Exception ex) {
+            l.onFailure(ex);
+        }
```
Why do we need the try-catch clause now? It wasn't present before. I guess that was a bug, where we didn't pass on exceptions from resolution to the listener? Is there a way to reproduce this issue?

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -303,97 +304,71 @@ public void analyzedPlan(
+            l -> enrichPolicyResolver.resolvePolicies(targetClusters, unresolvedPolicies, l)
+        ).<ListenerResult>andThen((l, enrichResolution) -> resolveFieldNames(parsed, enrichResolution, l));
+        // first resolve the lookup indices, then the main indices
+        // if (indices.size() > 1) {
+        // // Note: JOINs on more than one index are not yet supported
+        // listener.onFailure(new MappingException("More than one LOOKUP JOIN is not supported"));
```
Leftover?

Actually, I thought we had a check like this here a while ago. `preAnalysis.indices.size() <= 1` is still something we should assert, no? There can only be 1 main index (pattern) atm.

craigtaverner replied:
Actually I put this here temporarily as a fallback option if we, for some other reason, wanted to revert to single-index lookups. This is now the easiest place to do that check. My plan was, once all tests pass, remove this code. So I think it is now time to do that.

craigtaverner replied:
This came from the check in preAnalyzeLookupIndices, which no longer does that check. However, the preAnalyzeIndices method still does this check, and that is about multiple FROM commands. I think it likely that any future JOIN work will happen on top of lookup-join, so the check in preAnalyzeIndices should probably stay forever.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -303,97 +304,71 @@ public void analyzedPlan(
+        }).addListener(logicalPlanListener);
+    }
+
+    private void preAnalyzeLookupIndices(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
```
nit:
```suggestion
    private void preAnalyzeLookupIndex(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
```

craigtaverner replied:
Done in a separate commit, since accepting this suggestion here would cause compilation errors.

alex-spies reviewed (commented):
Thanks @craigtaverner , the capability to perform multiple LOOKUPs is great!

I still have to finish reviewing the changes to `EsqlSession` and the field names tests, but wanted to give you an early update with some additional test suggestions that we can already try.

alex-spies reviewed (approved):
Heya, the added tests are great, as is the ability to use multiple LOOKUP JOINs. Thanks @craigtaverner !

There's 2 things I'd ideally do before merging:
- Add a couple more test cases (see comments)
- Attempt a small tweak to the fieldNames method to request fewer fields from the main index in some cases (also in comments)

The latter should help avoid wasteful field caps calls in some cases, which I believe are the reason for having the `fieldNames` method in the first place.

But this is already in great shape as-is, so I'm gonna go and approve. If you think some of my remarks are more something for a follow-up, let's just add them as TODO comments to the code and move on.

My suggested additional csv test cases could also be unit tests in some case - let's use whichever test type you see fit, @craigtaverner !

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -303,97 +304,71 @@ public void analyzedPlan(
+
+    private void preAnalyzeLookupIndices(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
+        TableIdentifier table = tableInfo.id();
+        Set<String> fieldNames = listenerResult.wildcardJoinIndices().contains(table.index())
+            ? IndexResolver.ALL_FIELDS
```
@alex-spies 

> Maybe we should add a TODO comment here to remind us of the optimization potential.

Imo, TODOs are way less visible and we act way less on them compared to a **github issue**.

> Because we have only one set of fieldNames in the listenerResult, queries like FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3 either have to ask for fields field, f1, f2, f3 from both indices (slightly wasteful), or only for field from idx (correct) and * from lu_idx (quite wasteful).

**Be careful here**, we risk eliminating some edge cases. `fieldNames` should cover as many fields as possible (VS. using `*`), not as few as possible.

> FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3 either have to ask for fields `field, f1, f2, f3` from both indices (slightly wasteful)

This is not wasteful, this is the bare minimum. **This list cannot be smaller than this**. Whatever the user used in the query we need to ask field_caps to establish if the user didn't have a typo or wanted to use incompatible types fields.

> We ask for more fields than we need, and we can determine this already.

Nop. As I mentioned above, this is not a case of having the smallest list possible based on LogicalPlan tree analysis and optimization. The analysis and optimization happens after `_field_caps` call. You cannot establish which fields you need and which you don't **before** the field_caps call. `fieldNames` is somewhat blind to the logical meaning of a query, it is extracting the field names in a completely dumb way.

> I think it'd be optimal to have 1 set of field names per lookup index. This way, we can ask for fields f1, f2, f3 and field from lu_idx, and just for field from idx.

Same as above. It has the potential to be a dangerous over-optimization and, gut feeling-wise, I am against this.

> We perform the field caps requests one after the other instead of in parallel.

I agree, this should be investigated separately from this PR. Since we have this idea now, @craigtaverner please create a github issue so that we don't lose track of it. PR comments tends to be lost and forgotten after the PR is merged. This is less likely with a TODO, but a GH issue is the best reminder imo.

> When the same lookup join index is used in 2 lookup joins, we perform 2 field caps requests.

Indeed. Either in this PR, or in a new one this can be improved. GH issue if it will be a separate PR please.

alex-spies replied:
> This is not wasteful, this is the bare minimum. This list cannot be smaller than this.
This is about the query `FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3`.
The list of fields required from `idx` can be smaller; we can be certain that it is just `field`. It cannot be smaller for `lu_idx`.

In particular, for queries like `FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP *f*`, the current approach asks for `field` and `*f*` from `idx`, but it requires only `field`.

I agree that sophisticated pruning needs to be done later, in the logical plan. However, here we ask for fields and patterns, which we know are required only for lookup indices, from the main index. I recall that for certain queries a large amount of time is spent performing field caps calls, so it seemed to me that avoiding large field caps responses was important, no?

alex-spies replied:
To be clear, what I'm proposing is that we just don't mix the required fields for lookup indices into the required fields for the main index. That'd also come with a clearer separation of concerns. It should not affect, at all, the correctness of the field caps call made for the main index.

astefan replied:
Thanks for explaining further. This suggestion needs a completely separate issue, we need to think well about it and build it carefully. We cannot afford to miss a field from the list we send to `_field_caps`.

alex-spies replied:
@astefan and I had a quick chat offline. We concluded that it's the least risky to go with this PR's approach, but let's open a GH issue + leave a comment inside `fieldNames`. E.g. in lines 577 ff., to explain that adding the
`keepCommandReferences` to the `keepJoinReferences` ultimately makes them appear as part of the `fieldNames`, even though the field names added this way are only required from the respective lookup index, not from the main index.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -593,10 +571,16 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
+                    wildcardJoinIndices.add(((UnresolvedRelation) join.right()).table().index());
+                } else {
+                    // Keep commands can reference the join columns with names that shadow aliases, so we block their removal
+                    keepJoinReferences.addAll(keepCommandReferences);
```
This remark belongs to the same discussion as above, so let's just add a comment + GH issue as concluded [here](https://github.com/elastic/elasticsearch/pull/118429#discussion_r1882551480).

astefan commented:
Since this PR touches code in `EsqlSession` related to how listeners react and how they are used, I would suggest adding some tests in `RequestIndexFilteringTestCase`. This test class is about using a `filter` [in the request](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-rest.html#esql-rest-filtering) to see how `EsqlSession` code works when the query fails **with the filter on the _field_caps call** with an error of "Field not found" or similar and when the same query succeeds when `EsqlSession` code removes the filter from the **second** (retried) `_field_caps` call. Now, that I'm writing these lines, I am thinking this is also a future improvement in this area (cc @alex-spies) where we already have the fields for `lookup join` indices from the first field_caps call with filter and for the second call we don't need to resolve the `lookup` indices' fields anymore.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -631,6 +615,10 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
         });
         // Add JOIN ON column references afterward to avoid Alias removal
         references.addAll(keepJoinReferences);
+        // If any JOIN commands need wildcard field-caps calls, persist the index names
```
Is this comment ok for every JOIN command we'll have, or is something specific to `lookup join`.
Either way, I suggest being specific.

craigtaverner replied:
This is true as long as the LOOKUP JOIN command does not have a `WITH` clause, which is true at the moment for all joins, but will not be true at some future point. We can only fix this once that syntax exists. Without `WITH` we imply `*`.

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/session/IndexResolverFieldNamesTests.java:
```diff
@@ -1363,8 +1363,116 @@ public void testMetrics() {
+    public void testLookupJoin() {
+        assertFieldNames(
+            "FROM employees | KEEP languages | RENAME languages AS language_code | LOOKUP JOIN languages_lookup ON language_code",
+            Set.of("languages", "language_code", "language_code.*", "languages.*"),
```
Please, group these based on the root field to keep the style of the test class: `"languages", "languages.*", "language_code", "language_code.*"`

craigtaverner replied:
Done

craigtaverner committed:
Fix failing test and renamed method call

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
```diff
@@ -325,3 +360,150 @@ count:long | type:keyword
+left         | 172.21.0.5        | Connected to 10.1.0.1 | right         | Development | Success
+;
+
+lookupIPAndMessageFromRowWithShadowingKeep-Ignore
```
Un-ignored, after fixing the test, which had a mistake in the KEEP command.

craigtaverner committed:
Added more csv-spec tests with more KEEPS and STATS before LOOKUP JOIN

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
```diff
@@ -325,3 +360,150 @@ count:long | type:keyword
+lookupIPAndMessageFromIndexStatsKeep
+required_capability: join_lookup_v5
+
+FROM sample_data
+| EVAL client_ip = client_ip::keyword
+| LOOKUP JOIN clientips_lookup ON client_ip
+| LOOKUP JOIN message_types_lookup ON message
+| KEEP client_ip, env, type
+| STATS count = count(*) BY env, type
+| SORT count DESC, env ASC, type ASC
```
csv-spec tests added, and they pass

craigtaverner committed:
Merge branch 'fix-lookupjoin-wildcard-fieldcaps' of github.com:craigtaverner/elasticsearch into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Increment capability number to prevent overlap with other work

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Fix compile error after merging main

craigtaverner committed:
Merge branch 'fix-lookupjoin-wildcard-fieldcaps' of github.com:craigtaverner/elasticsearch into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Added unit tests and csv-spec tests for lookup-join on the same index twice

craigtaverner committed:
Rename class for clarity of purpose

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -703,22 +691,36 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
     }
 
-    private record ListenerResult(
+    record ListenerResult(
```
Done

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -703,22 +691,36 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
+        Map<String, IndexResolution> lookupIndices,
         EnrichResolution enrichResolution,
-        Set<String> fieldNames
+        Set<String> fieldNames,
```
No, this contains fields that might come from the main or the lookup indices. For some queries we cannot determine where the fields came from, so need to perform field-caps calls on each index for the complete set, after which we will know which indices each field comes from.

alex-spies replied:
Sorry, you're right, I got confused at the time I reviewed this.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -554,10 +517,24 @@ private static void analyzeAndMaybeRetry(
+    private static void resolveFieldNames(LogicalPlan parsed, EnrichResolution enrichResolution, ActionListener<ListenerResult> l) {
+        try {
+            // we need the match_fields names from enrich policies and THEN, with an updated list of fields, we call field_caps API
+            var enrichMatchFields = enrichResolution.resolvedEnrichPolicies()
+                .stream()
+                .map(ResolvedEnrichPolicy::matchField)
+                .collect(Collectors.toSet());
```
I was trying to keep the changes to fieldNames as small a localized as possible, to make reviews easier, and reduce potential conflicts.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -593,10 +571,16 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
                 if (join.config().type() instanceof JoinTypes.UsingJoinType usingJoinType) {
                     keepJoinReferences.addAll(usingJoinType.columns());
                 }
+                if (keepCommandReferences.isEmpty()) {
+                    // No KEEP commands after the JOIN, so we need to mark this index for "*" field resolution
+                    wildcardJoinIndices.add(((UnresolvedRelation) join.right()).table().index());
+                } else {
+                    // Keep commands can reference the join columns with names that shadow aliases, so we block their removal
+                    keepJoinReferences.addAll(keepCommandReferences);
```
Added both unit tests and csv-spec tests for multiple joins on the same lookup index, and it all works as expected. The rules for each index is that if there is a KEEP after that index, it uses fieldNames, otherwise it needs to use `*`, regardless of how many times that index is joined with or if any previous joins had a KEEP afterwards.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -554,10 +517,24 @@ private static void analyzeAndMaybeRetry(
+        } catch (Exception ex) {
+            l.onFailure(ex);
+        }
+    }
```
This code was there before, in the join.indices.size==0 case, and was specifically wrapping the listener.onResponse call. All other places where we call listener.onResponse we have a try/catch, so I think we should keep it. The one time we did see something being thrown and caught here was when one PR added behaviour to PlanNode.output() that relied on fields being resolved, but fieldNames() actually calls down to this, and fails if the fields are not yet resolved. This was fixed a couple of weeks back, but I can imagine future scenarios where this could happen.

craigtaverner committed:
Added tests for wildcard KEEP

craigtaverner committed:
Reordered expected fields to be more like rest of test class

craigtaverner committed:
Added tests using results from one lookup as join key in next lookup

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/lookup-join.csv-spec:
Added a lot more tests now, even using results from one lookup as join key for the next lookup

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/session/IndexResolverFieldNamesTests.java:
Added unit tests with wildcard KEEP.

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Added RequestIndexFilteringTest on missing LOOKUP JOIN index

craigtaverner committed:
Block running in mixed clusters

astefan reviewed (commented):
@craigtaverner I was hoping with [this comment](https://github.com/elastic/elasticsearch/pull/118429/files#r1881862228) from Alex you would catch this one.

I've taken that exact query that Alex added a review to and I simply moved the `keep` command and placed it between the two `lookup`s (like how he suggested) with this final query:

```
ROW left = "left", client_ip = "172.21.0.5", message = "Connected to 10.1.0.1", right = "right"
| LOOKUP JOIN clientips_lookup ON client_ip
| KEEP left, client_ip, message, right
| LOOKUP JOIN message_types_lookup ON message
```

It fails with 

```
path: /_query, params: {format=txt, error_trace=true}, status: 500 java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 0
        at org.elasticsearch.compute.operator.lookup.MergePositionsOperator.<init>(MergePositionsOperator.java:76)
        at org.elasticsearch.xpack.esql.enrich.AbstractLookupService.doLookup(AbstractLookupService.java:329)
        at org.elasticsearch.xpack.esql.enrich.AbstractLookupService$TransportHandler.messageReceived(AbstractLookupService.java:447)
        at org.elasticsearch.xpack.esql.enrich.AbstractLookupService$TransportHandler.messageReceived(AbstractLookupService.java:442)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:90)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.transport.TransportService$6.doRun(TransportService.java:1098)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1575)
```

If it's something easy to fix, do it here, otherwise merge this PR and address this problem ^ in a separate PR right away.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Added more tests for KEEP before, between and after joins when using ROW

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

astefan commented:
https://github.com/elastic/elasticsearch/issues/118778

astefan reviewed (approved):
LGTM

craigtaverner committed:
Fixed capability name after merging main

craigtaverner committed:
Fix failing tests after merging in main

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Just use one index name

alex-spies reviewed (approved):
Reviewed the changes to LogicalPlanOptimizerTests, these LGTM! Thanks Craig!

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

craigtaverner committed:
Merge branch 'main' into fix-lookupjoin-wildcard-fieldcaps

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/118799)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/118799">](https://github.com/elastic/elasticsearch/pull/118799) |
